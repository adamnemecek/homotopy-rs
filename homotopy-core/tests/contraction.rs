use homotopy_core::*;
use insta::*;

#[test]
fn scalar() {
    let x = Diagram::from(Generator::new(0, 0));
    let s = DiagramN::new(Generator::new(1, 2), x.identity(), x.identity()).unwrap();
    let t = DiagramN::new(Generator::new(2, 2), x.identity(), x.identity()).unwrap();
    let diagram = s.attach(t.clone(), Boundary::Target, &[]).unwrap();

    assert!(diagram.contract(&[], 0, None).is_none());
    assert_debug_snapshot!(
        "scalar_biased_left",
        diagram.contract(&[], 0, Some(Bias::Lower)).unwrap()
    );
    assert_debug_snapshot!(
        "scalar_biased_right",
        diagram.contract(&[], 0, Some(Bias::Higher)).unwrap()
    );
}
