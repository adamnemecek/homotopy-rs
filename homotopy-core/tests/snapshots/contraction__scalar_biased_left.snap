---
source: homotopy-core/tests/contraction.rs
expression: "diagram.contract(&[], 0, Some(Bias::Lower)).unwrap()"
---
DiagramN {
    source: DiagramN {
        source: Diagram0(
            0:0,
        ),
        cospans: [],
    },
    cospans: [
        Cospan {
            forward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        source: [],
                        target: Cospan {
                            forward: Rewrite0(
                                0:0,
                                1:2,
                            ),
                            backward: Rewrite0(
                                0:0,
                                1:2,
                            ),
                        },
                        slices: [],
                    },
                    Cone {
                        index: 0,
                        source: [],
                        target: Cospan {
                            forward: Rewrite0(
                                0:0,
                                2:2,
                            ),
                            backward: Rewrite0(
                                0:0,
                                2:2,
                            ),
                        },
                        slices: [],
                    },
                ],
            },
            backward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        source: [],
                        target: Cospan {
                            forward: Rewrite0(
                                0:0,
                                1:2,
                            ),
                            backward: Rewrite0(
                                0:0,
                                1:2,
                            ),
                        },
                        slices: [],
                    },
                    Cone {
                        index: 0,
                        source: [],
                        target: Cospan {
                            forward: Rewrite0(
                                0:0,
                                2:2,
                            ),
                            backward: Rewrite0(
                                0:0,
                                2:2,
                            ),
                        },
                        slices: [],
                    },
                ],
            },
        },
    ],
}
